src/Facet/Unify.hs

```diff
@@ -61,59 +67,41 @@ unifyType = curry $ \case
+flexFlex :: (HasCallStack, Has (Reader ElabContext :+: Reader StaticContext :+: State (Subst Type) :+: Throw Err :+: Throw (WithCallStack UnifyErrReason) :+: Writer Usage) sig m) => Meta -> Meta -> m Type
 flexFlex v1 v2
   | v1 == v2  = pure (metavar v2)
   | otherwise = gets (\ s -> (lookupMeta v1 s, lookupMeta v2 s)) >>= \case
-    (Just t1, Just t2) -> unifyType (tm t1) (tm t2)
-    (Just t1, Nothing) -> unifyType (metavar v2) (tm t1)
-    (Nothing, Just t2) -> unifyType (metavar v1) (tm t2)
+    (Just t1, Just t2) -> unifyType t1 t2
+    (Just t1, Nothing) -> unifyType (metavar v2) t1
+    (Nothing, Just t2) -> unifyType (metavar v1) t2
     (Nothing, Nothing) -> solve v1 (metavar v2)
 
-solve :: (HasCallStack, Has (Reader ElabContext :+: Reader StaticContext :+: Reader (Exp Subject :=: Act Subject) :+: State Subst :+: Throw Err :+: Writer Usage) sig m) => Meta -> Type -> m Type
+solve :: (HasCallStack, Has (Reader ElabContext :+: Reader StaticContext :+: State (Subst Type) :+: Throw Err :+: Throw (WithCallStack UnifyErrReason) :+: Writer Usage) sig m) => Meta -> Type -> m Type
 solve v t = do
   d <- depth
-  if occursIn (== Free (Left v)) d t then
+  if occursIn v d t then
     occurs v t
   else
     gets (lookupMeta v) >>= \case
-      Nothing          -> t <$ modify (solveMeta v t)
-      Just (t' ::: _T) -> unifyType t' t >>= \ t'' -> t'' <$ modify (solveMeta v t'')
-
-
--- Equating
-
-runEq :: Applicative m => EmptyC m () -> m Bool
-runEq = execEmpty
-
-eqType :: (HasCallStack, Has (Empty :+: Reader ElabContext :+: Reader StaticContext :+: State Subst :+: Throw Err :+: Writer Usage) sig m) => Type -> Type -> m ()
-eqType = curry $ \case
-  (VComp s1 t1, VComp s2 t2)           -> eqSpine eqInterface s1 s2 *> eqType t1 t2
-  (VComp _ t1, t2)                     -> eqType t1 t2
-  (t1, VComp _ t2)                     -> eqType t1 t2
-  (VForAll _ t1 b1, VForAll n t2 b2)   -> depth >>= \ d -> guard (t1 == t2) *> (Binding n zero (Left t2) |- eqType (b1 (free d)) (b2 (free d)))
-  (VForAll{}, _)                       -> empty
-  (VArrow _ _ a1 b1, VArrow _ _ a2 b2) -> eqType a1 a2 *> eqType b1 b2
-  (VArrow{}, _)                        -> empty
-  (VNe v1 sp1, VNe v2 sp2)             -> guard (v1 == v2) *> eqSpine eqType sp1 sp2
-  (VNe{}, _)                           -> empty
-  (VString, VString)                   -> pure ()
-  (VString, _)                         -> empty
-
-eqInterface :: (HasCallStack, Has (Empty :+: Reader ElabContext :+: Reader StaticContext :+: State Subst :+: Throw Err :+: Writer Usage) sig m) => Interface Type -> Interface Type -> m ()
-eqInterface (Interface h1 sp1) (Interface h2 sp2) = guard (h1 == h2) *> eqSpine eqType sp1 sp2
-
-eqSpine :: (Foldable t, Zip t, Has Empty sig m) => (a -> b -> m ()) -> t a -> t b -> m ()
-eqSpine f sp1 sp2 = guard (length sp1 == length sp2) *> zipWithM_ f sp1 sp2
```

Comment:
Instead of equating, we now unify, but return in `Maybe` if an error was thrown instead of failing outright. Thus we can still search for interfaces in the signature stack, but don’t require an entirely separate code path to do so, which meant having two equalities (that could fall out of sync) instead of only one.